ext/opcache/jit/zend_jit.c

```diff
@@ -2656,6 +2696,198 @@ static zend_lifetime_interval** zend_jit_allocate_registers(const zend_op_array
 	return NULL;
 }
 
+#else /* ZEND_JIT_IR */
+
+static void zend_jit_allocate_registers(zend_jit_ctx *ctx, const zend_op_array *op_array, zend_ssa *ssa)
+{
+	void *checkpoint;
+	int candidates_count, i;
+	zend_jit_reg_var *ra;
+
+	checkpoint = zend_arena_checkpoint(CG(arena));
+	ra = zend_arena_calloc(&CG(arena), ssa->vars_count, sizeof(zend_jit_reg_var));
+	candidates_count = 0;
+	for (i = 0; i < ssa->vars_count; i++) {
+		if (zend_jit_may_be_in_reg(op_array, ssa, i)) {
+			ra[i].ref = IR_NULL;
+			candidates_count++;
+		}
+	}
+	if (!candidates_count) {
+		zend_arena_release(&CG(arena), checkpoint);
+		return;
+	}
+
+	if (JIT_G(opt_flags) & ZEND_JIT_REG_ALLOC_GLOBAL) {
+		/* Naive SSA resolution */
+		for (i = 0; i < ssa->vars_count; i++) {
+			if (ssa->vars[i].definition_phi && !ssa->vars[i].no_val) {
+				zend_ssa_phi *phi = ssa->vars[i].definition_phi;
+				int k, src;
+
+				if (phi->pi >= 0) {
+					src = phi->sources[0];
+					if (ra[i].ref) {
+						if (!ra[src].ref) {
+							ra[i].flags |= ZREG_LOAD;
+						} else {
+							ra[i].flags |= ZREG_PI;
+						}
+					} else if (ra[src].ref) {
+						ra[src].flags |= ZREG_STORE;
+					}
+				} else {
+					int need_move = 0;
+
+					for (k = 0; k < ssa->cfg.blocks[phi->block].predecessors_count; k++) {
+						src = phi->sources[k];
+						if (src >= 0) {
+							if (ssa->vars[src].definition_phi
+							 && ssa->vars[src].definition_phi->pi >= 0
+							 && phi->block == ssa->vars[src].definition_phi->block) {
+								/* Skip zero-length interval for Pi variable */
+								src = ssa->vars[src].definition_phi->sources[0];
+							}
+							if (ra[i].ref) {
+								if (!ra[src].ref) {
+									need_move = 1;
+								}
+							} else if (ra[src].ref) {
+								need_move = 1;
+							}
+						}
+					}
+					if (need_move) {
+						if (ra[i].ref) {
+							ra[i].flags |= ZREG_LOAD;
+						}
+						for (k = 0; k < ssa->cfg.blocks[phi->block].predecessors_count; k++) {
+							src = phi->sources[k];
+							if (src >= 0) {
+								if (ssa->vars[src].definition_phi
+								 && ssa->vars[src].definition_phi->pi >= 0
+								 && phi->block == ssa->vars[src].definition_phi->block) {
+									/* Skip zero-length interval for Pi variable */
+									src = ssa->vars[src].definition_phi->sources[0];
+								}
+								if (ra[src].ref) {
+									ra[src].flags |= ZREG_STORE;
+								}
+							}
+						}
+					} else {
+						ra[i].flags |= ZREG_PHI;
+					}
+				}
+			}
+		}
+
+		/* Remove useless register allocation */
+		for (i = 0; i < ssa->vars_count; i++) {
+			if (ra[i].ref &&
+			    ((ra[i].flags & ZREG_LOAD) ||
```

Comment:
From above, if `ra[i].ref` is true, then `ra[i].flags` will always be `ZREG_LOAD`, maybe? Actually, not always. Looks like it's conditional, but leaving this comment here for now.

Reply:
The idea of this loop is avoiding register allocation for variables that are not used directly but have to be loaded and stored at the same memory location.